### DISCUSSION TITLE
Form mappings in Scala3 - CaseClass.unapply no longer usable?

### DISCUSSION THREAD

**Discussion by jxtps** [playframework/playframework#12292] [Q&A] [answered]
I'm trying to migrate a Play 3.0.1 app from Scala 2.13.12 to Scala 3.3.1, but am running into a problem with my forms. The Scala3 compiler complains that my `CaseClass.unapply` methods are not applicable as the third argument to the `mapping` function: 
```
  case class MyError(code: Int, message: String)
  val MyErrorMapping: Mapping[MyError] = mapping(
    "code" -> number,
    "message" -> nonEmptyText
  )(MyError.apply)(MyError.unapply)  // <-- Works in Scala2, but not in Scala3!?
```

On https://www.playframework.com/documentation/3.0.x/ScalaForms the example provided adds a custom `unapply` method: 

```
case class UserData(name: String, age: Int)
object UserData {
  def unapply(u: UserData): Option[(String, Int)] = Some(u.name, u.age)
}
```

Do I have to manually specify each and every `unapply` method to be able to migrate to Scala3? 

Thanks!

**Comment by jxtps** [comment]
Ok, so this is not looking good: https://docs.scala-lang.org/scala3/guides/migration/incompat-other-changes.html#explicit-call-to-unapply

Basically the `unapply` method has changed its signature: 
```
object MyError {
  def unapply(myError: MyError): Option[(Int, String)] = ... // <- Scala 2
  def unapply(myError: MyError): MyError= myError // <- Scala 3
}
```
The Scala2 unapply method is quite useful for the Play forms, but the Scala 3 unapply method obviously is not. 

Not sure how to proceed here - adding hand-cranked unapply methods is not a palatable path forward.

**Reply by jtjeferreira** [reply, marked as answer]
Any options described in this [thread](https://github.com/lampepfl/dotty/issues/2335#issuecomment-1309916758) help?

**Reply by jxtps** [reply]
I haven't tested it, but to save ppl a click, here's the proposed solution: 
```scala
  def unapply[P <: Product](p: P)(using m: scala.deriving.Mirror.ProductOf[P]): Option[m.MirroredElemTypes] =
    Some(Tuple.fromProductTyped(p))
```
